man/fi_peer.3.md

```diff
@@ -192,6 +192,120 @@ on the local node.
+    int (*get_tag_entry)(struct fid_peer_srx *srx,
+        struct fi_peer_rx_entry *entry);
+};
+
+struct fi_ops_srx_peer {
+    size_t size;
+    int (*start_msg)(struct fid_peer_srx *srx,
+        struct fi_peer_rx_entry *entry);
+    int (*start_tag)(struct fid_peer_srx *srx,
+        struct fi_peer_rx_entry *entry);
+    int (*discard_msg)(struct fid_peer_srx *srx,
+        struct fi_peer_rx_entry *entry);
+    int (*discard_tag)(struct fid_peer_srx *srx,
+        struct fi_peer_rx_entry *entry);
+};
+
+struct fid_peer_srx {
+    struct fid fid;
+    struct fi_ops_srx_owner *owner_ops;
+    struct fi_ops_srx_peer *peer_ops;
+};
+
+struct fi_peer_srx_context {
+    size_t size;
+    struct fid_peer_srx *srx;
+};
+```
+The ownership of structure field values and callback functions is similar
+to those defined for peer CQs, relative to owner versus peer ops.
+
+## fi_ops_srx_owner::get_msg_entry() / get_tag_entry()
+
+These calls are invoked by the peer provider to obtain the receive buffer(s)
+where an incoming message should be placed.  The fi_peer_rx_entry is allocated
+by the peer provider and initialized with information about the received
+message.  If source addressing is required, the addr field will specify the
+source address for the incoming message, otherwise, the source address will
+be set to FI_ADDR_NOTAVAIL.  The size field indicates the received message
+size.  This field is used by the owner when handling multi-receive data
+buffers, but may be ignored otherwise.  The peer provider is responsible
+for checking that an incoming message fits within the provided buffer space.
+The tag field is set for tagged messages.
+(TBD: The peer may need to update the src addr if the remote endpoint is
+inserted into the AV after the message has been received.)
+
+The next, prev, and context fields may be used by the owner while the
+fi_peer_rx_entry is under its control.  On input to the get entry calls,
+the iov_cnt should be set by the peer to the number of entries in the iov
+array.  When returning the fi_peer_rx_entry to the peer, the owner will
+update the iov_cnt to the number of iov entries containing valid data.
+The iov entries will reference the receive buffer(s) where the message
+should be placed.
+
+The get_msg_entry() and get_tag_entry() calls may complete synchronously
+or asynchronously, depending on whether an application buffer is ready to
+receive the incoming data.  A return value of FI_SUCCESS indicates that the
+call has completed synchronously.  In this case, the fi_peer_rx_entry is
+returned to the peer with the iov referencing the receive buffer(s) and
+the context field set to the application's context (to pass to any
+completion).  A return value of -FI_EINPROGRESS indicates that a matching
+receive entry was not found and the request has been queued.  The owner
+will notify the peer when a match has been made or if the message should
+be discarded.
+
+## fi_ops_srx_peer::start_msg() / start_tag()
+
```

Comment:
If I understand correctly, these two functions are implemented by peer and are called by owner. Right?

Reply:
Correct.  The owner would call these in response to the application posting a matching receive.

Reply:
So the workflow is:

1. application call `fi_recv/fi_trecv` on owner.
2. owner save the receive buffer in a queue
3. Meanwhile, peer receive a message, and need to find a receive buffer for it.
4. Peer call `owner->get_msg_entry/get_tag_entry`.
5. If there are receive in queue, `get_msg_entry` will return the buffer, and return 0. Otherwise, `get_msg/tag_entry` will put the info in an unexpected msg list, and return -FI_EINPROGRESS.
6. when application put in new receive, it first get matched with unexpected msg list, and owner call `peer->start_msg`

I guess a graph would be helpful.

Reply:
`start_msg` is not easy to understand. May be `start_recv_msg`?

Also, can we put the `peer_rx_entry` as argument? since owner already has that information.

Reply:
The flow is correct.  This is an operation of the shared receive context -- 'recv' is a given.  The peer_rx_entry is an argument.  The arguments are just not listed as part of the heading in the man page.